src/effects/VST3/internal/ParameterValueQueue.cpp

```diff
@@ -0,0 +1,90 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file ParameterValueQueue.cpp
+
+  @author Vitaly Sverchinsky
+
+  @brief Part of Audacity VST3 module
+
+**********************************************************************/
+
+#include "ParameterValueQueue.h"
+
+#include <wx/debug.h>
+
+internal::ParameterValueQueue::ParameterValueQueue()
+{
+   FUNKNOWN_CTOR
+}
+
+void internal::ParameterValueQueue::initialize(Steinberg::Vst::ParamID id)
+{
+   wxASSERT(!mInitialized);//misuse
+
+   mInitialized = true;
+   mId = id;
+}
+
+bool internal::ParameterValueQueue::isInitialized() const noexcept
+{
+   return mInitialized;
+}
+
+void internal::ParameterValueQueue::dispose() noexcept
+{
+   mInitialized = false;
+   mPoints.clear();
+}
+
+Steinberg::Vst::ParamID internal::ParameterValueQueue::getParameterId()
+{
+   wxASSERT(mInitialized);
+   return mId;
+}
+
+Steinberg::int32 internal::ParameterValueQueue::getPointCount()
+{
+   wxASSERT(mInitialized);
+   return static_cast<int>(mPoints.size());
+}
+
+Steinberg::tresult internal::ParameterValueQueue::getPoint(Steinberg::int32 index, Steinberg::int32& sampleOffset,
+   Steinberg::Vst::ParamValue& value)
+{
+   wxASSERT(mInitialized);
+   if(index >= 0 && index < mPoints.size())
+   {
+      sampleOffset = mPoints[index].sampleOffset;
+      value = mPoints[index].paramValue;
+      return Steinberg::kResultOk;
+   }
+   return Steinberg::kInvalidArgument;
+}
+
+Steinberg::tresult internal::ParameterValueQueue::addPoint(Steinberg::int32 sampleOffset,
+   Steinberg::Vst::ParamValue value, Steinberg::int32& index)
```

Comment:
This function never assigns `index`.  But does the caller expect that you do?

Reply:
Our implementation does not use it. VST documentation is not very wordy about the use of this parameter, and there are no examples of how this parameter is used in the SDK samples as well. Moreover, it's not stated that the index should be persistent, but plugins can access our implementation through `IParameterValueQueue ` interface. So yes, it would be safer to assign a value to it.